---
title: "epigenomics"
author: "Sofia Salazar & Bruno Villalobos"
date: "3/24/2022"
output: html_document
---

#### Blueprint pipeline

Starting with fastq file from ChIP-seq experiment of H3K4me3 in erytrhoblasts. Mapping was performed with bwa 0.7.7 to the reference genome GRCh38. BAM files are sorted and duplicates were marked using picard. Data is filetered discarding unmapped reads and those with mapping quality less than 5 using samtools 0.1.19 and then re-filtered to remove PCR or optical duplicate reads. Fragment size is then modelled using a PhantomPeakQualTools R script. Peaks were called using MACS2 (2.0.10.20131216) using predicted fragment sizes and avoiding the -broad flag for H3K4me3 marks. bigWigs are produced using align2RawSignal using the fragment size predicted by PhantomPeakQualTools.

#### Histone enrichment analysis

We chose H3K4me3 because, since we are analysing these marks in erythrobalsts cells, which are precursor cells of erythrocytes, it would be interesting to see the landscape of the active genes in these pre-state cells. Since H3K4me3 is found in the 5' region of genes and is associated with transcriptional regions, then the analysis of these marks can be a useful tool for out landscape visualization.

#### Data download and processing

Downloading list of all interactions detected between active promoters and enhancers.

```{bash}
qlogin
cd /mnt/Timina/bioinfoII/ssalazar/epigenomics/data
wget https://www.cell.com/cms/10.1016/j.cell.2016.09.037/attachment/5bc79f6f-1b69-4192-8cb8-4247cc2e0f39/mmc4.zip
```


Downloading signal files (bigWig) from bluePrint
```{bash}
wget http://ftp.ebi.ac.uk/pub/databases/blueprint/data/homo_sapiens/GRCh38/cord_blood/S002R5/erythroblast/ChIP-Seq/NCMLS/S002R5H1.ERX300721.H3K4me3.bwa.GRCh38.20150528.bw

wget http://ftp.ebi.ac.uk/pub/databases/blueprint/data/homo_sapiens/GRCh38/cord_blood/S002S3/erythroblast/ChIP-Seq/NCMLS/S002S3H1.ERX300734.H3K4me3.bwa.GRCh38.20150528.bw

wget http://ftp.ebi.ac.uk/pub/databases/blueprint/data/homo_sapiens/GRCh38/cord_blood/S002R5/erythroblast/ChIP-Seq/NCMLS/S002R5H1.ERX337057.Input.bwa.GRCh38.20150528.bw

wget http://ftp.ebi.ac.uk/pub/databases/blueprint/data/homo_sapiens/GRCh38/cord_blood/S002S3/erythroblast/ChIP-Seq/NCMLS/S002S3H1.ERX337056.Input.bwa.GRCh38.20150528.bw
```

Filtering the interaction list and getting two different bed files, one for promoters and one for enhancers
```{bash}
grep "Ery" ActivePromoterEnhancerLinks.tsv | awk '{ print $1, $2, $3 }' | sed -e 's/ /\t/g' > promotersEry.bed

grep "Ery" ActivePromoterEnhancerLinks.tsv | awk '{ print $5, $6, $7 }' | sed -e 's/ /\t/g' > enhancersEry.bed
```

We need to change the positions on the bed files since they are based on the GRCh37 reference genome version according to the article, and the bigWig files from Blueprint are based on the GRCh38 reference genome. 

We first download the liftover chains to go from Hg19 to Hg38

```{bash eval = F}
wget https://hgdownload.cse.ucsc.edu/goldenpath/hg19/liftOver/hg19ToHg38.over.chain.gz
```

Then we use the `liftOver` tool from the UCSC module to do the conversion.

```{bash eval = F}
module load UCSC-executables/ucsc
liftOver promotersEry.bed hg19ToHg38.over.chain.gz promoters38.bed notpassed-promoters38Ery.bed
liftOver enhancersEry.bed hg19ToHg38.over.chain.gz enhancers38.bed notpassed-enhancers38Ery.bed
```

We get the regions that could be converted in files `promoters38.bed` and `enhancers38.bed` and the regions that had issues passed from one genome to another in the `notpassed-promoters38Ery.bed` and `notpassed-enhancers38Ery.bed` files. For both of the not passed files we have only two regions.


#### Is there an enrichment for H3K4me3 histone in the promoters or in the enhancers?

```{bash}
mv S002R5H1.ERX300721.H3K4me3.bwa.GRCh38.20150528.bw H3K4me3-male.bw
mv S002S3H1.ERX300734.H3K4me3.bwa.GRCh38.20150528.bw H3K4me3-female.bw
mv S002R5H1.ERX337057.Input.bwa.GRCh38.20150528.bw input-male.bw
mv S002S3H1.ERX337056.Input.bwa.GRCh38.20150528.bw input-female.bw
```

```{bash}
bigWigMerge H3K4me3-male.bw H3K4me3-female.bw H3K4me3-merged.bedGraph
bigWigMerge input-male.bw input-female.bw input-merged.bedGraph
```

```{bash}
wget https://hgdownload.cse.ucsc.edu/goldenpath/hg38/bigZips/hg38.chrom.sizes
awk 'NR!=1' H3K4me3-merged.bedGraph > headless-H3K4me3-merged.bedGraph
sort -k1,1 -k2,2n headless-H3K4me3-merged.bedGraph | awk '{print $1,$2,$3,$4}' OFS="\t" > H3K4me3-sorted.bedGraph
bedGraphToBigWig H3K4me3-sorted.bedGraph hg38.chrom.sizes H3K4me3.bw

awk 'NR!=1' input-merged.bedGraph > headless-input-merged.bedGraph
sort -k1,1 -k2,2n headless-input-merged.bedGraph | awk '{print $1,$2,$3,$4}' OFS="\t" > input-sorted.bedGraph
bedGraphToBigWig input-sorted.bedGraph hg38.chrom.sizes input.bw
```

Reduce noise of bigwigs with input files
```{bash}
# bigwigCompare -b1 H3K4me3-merged.bw -b2 input-merged.bw -o noiselessH3K4me3.bw –outFileFormat “bigwig”
```

https://www.nature.com/articles/nature19362
